cherab/core/atomic/zeeman.pyx

```diff
@@ -0,0 +1,124 @@
+                           of individual :math:`\pi`-polarised Zeeman components for a given
+                           magnetic field strength. The sum of all ratios must be equal to 1.
+    :param list wavelengths_sigma: A list of Function1D objects that provide the wavelengths
+                                   of individual :math:`\sigma`-polarised Zeeman components fo
+                                   a given magnetic field strength.
+    :param list ratios_sigma: A list of Function1D objects that provide the ratios
+                              of individual :math:`\sigma`-polarised Zeeman components for a given
+                              magnetic field strength. The sum of all ratios must be equal to 1.
+    """
+
+    def __init__(self, wavelengths_pi, ratios_pi, wavelengths_sigma, ratios_sigma):
+
+        if len(wavelengths_pi) != len(ratios_pi):
+            raise ValueError('The lengths of "wavelengths_pi" ({}) and "ratios_pi" ({}) do not match.'.format(len(wavelengths_pi),
+                                                                                                              len(ratios_pi)))
+
+        if len(wavelengths_sigma) != len(ratios_sigma):
+            raise ValueError('The lengths of "wavelengths_sigma" ({}) and "ratios_sigma" ({}) do not match.'.format(len(wavelengths_sigma),
+                                                                                                                    len(ratios_sigma)))
+
+        self._number_of_pi_lines = len(wavelengths_pi)
+        self._number_of_sigma_lines = len(wavelengths_sigma)
+
+        self._wavelengths = wavelengths_pi + wavelengths_sigma
+        self._ratios = ratios_pi + ratios_sigma
+
+        for i in range(len(self._wavelengths)):
+            self._wavelengths[i] = autowrap_function1d(self._wavelengths[i])
+            self._ratios[i] = autowrap_function1d(self._ratios[i])
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.initializedcheck(False)
+    @cython.cdivision(True)
+    cdef double[:, :] evaluate(self, double b, bint polarisation):
+
+        cdef int i, start, number_of_lines
+        cdef np.npy_intp multiplet_shape[2]
+        cdef double ratio_sum
+        cdef np.ndarray multiplet
+        cdef double[:, :] multiplet_mv
+        cdef Function1D wavelength, ratio
+
+        b = max(0, b)
+
+        if polarisation == PI_POLARISATION:
+            start = 0
+            number_of_lines = self._number_of_pi_lines
+        else:
+            start = self._number_of_pi_lines
+            number_of_lines = self._number_of_sigma_lines
+
+        multiplet_shape[0] = 2
+        multiplet_shape[1] = number_of_lines
+        multiplet = np.PyArray_SimpleNew(2, multiplet_shape, np.NPY_FLOAT64)
+        multiplet_mv = multiplet
+
+        ratio_sum = 0
+        for i in range(number_of_lines):
+            wavelength = self._wavelengths[start + i]
```

Comment:
Why would you use indexing here? Just select a pi or sigma list in the if statement above. Most of this code in unnecessary if you just dynamically choose which list you are processing:

```
if polarisation == PI_POLARISATION:
    data = self._pi_data
else:
    data = self._sigma_data
````

Also, len() is fast for lists in cython, there is no need to cache the value. You are falling for the trap of over optimising vs code clarity.

Reply:
Done.